### DISCUSSION TITLE
Sequencing of Kernels / Graphs?

### DISCUSSION THREAD

**Discussion by apowers313** [inducer/pycuda#458] [Q&A]
Hey, just learning CUDA and I'm loving PyCUDA.

I have a number of small kernels that pass data to each other and I don't want to use the host / CPU to coordinate all the kernel launching and data passing because of the overhead. I think that graphs are the solution that I need, but it looks like that [PR](https://github.com/inducer/pycuda/pull/424) is still in progress.

I started looking at the [CUDA Python Graph Management API](https://nvidia.github.io/cuda-python/module/cuda.html#graph-management), but CUDA Python is like writing in assembly and it's going to take forever to get that up and running. (Again, thanks for PyCUDA)

Is there some workaround for sequencing kernels and memory operations in PyCUDA? A way of calling CUDA C++ APIs directly or something?

Or is there some way having a kernel block on an event from another kernel, so that it can block until a kernel is done running?

Thanks for any insights. :)

**Comment by mitkotak** [comment]
Hey sorry for the stalled progress on that PR. If you have some bandwidth, I can guide you to have something working for your use case.

**Comment by apowers313** [comment]
Thanks for the quick reply. Yea, that'd be great.

**Comment by apowers313** [comment]
If some can show me how PyCUDA does it's C++ bindings I think I can put together a PR for building out some graph functionality.

**Reply by inducer** [reply]
https://gitlab.tiker.net/kaushikcfd/pycuda/-/merge_requests/2/diffs contains some work towards this, as does #343. What mostly needs to happen is that either version (or a combination of them) needs to be polished up enough to be mergeable.
